modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtGetFuture.java

```diff
@@ -259,77 +267,14 @@ public GridCacheVersion version() {
-        Map<KeyCacheObject, Boolean> mappedKeys = null;
-
-        // Assign keys to primary nodes.
-        for (Map.Entry<KeyCacheObject, Boolean> key : keys.entrySet()) {
-            int part = cctx.affinity().partition(key.getKey());
-
-            if (retries == null || !retries.contains(part)) {
-                if (!map(key.getKey(), forceKeys)) {
-                    if (retries == null)
-                        retries = new HashSet<>();
-
-                    retries.add(part);
-
-                    if (mappedKeys == null) {
-                        mappedKeys = U.newLinkedHashMap(keys.size());
-
-                        for (Map.Entry<KeyCacheObject, Boolean> key1 : keys.entrySet()) {
-                            if (key1.getKey() == key.getKey())
-                                break;
-
-                            mappedKeys.put(key.getKey(), key1.getValue());
-                        }
-                    }
-                }
-                else if (mappedKeys != null)
-                    mappedKeys.put(key.getKey(), key.getValue());
-            }
-        }
-
-        // Add new future.
-        IgniteInternalFuture<Collection<GridCacheEntryInfo>> fut = getAsync(mappedKeys == null ? keys : mappedKeys);
-
-        // Optimization to avoid going through compound future,
-        // if getAsync() has been completed and no other futures added to this
-        // compound future.
-        if (fut.isDone() && !hasFutures()) {
-            if (fut.error() != null)
-                onDone(fut.error());
-            else
-                onDone(fut.result());
-
-            return;
-        }
-
-        add(fut);
-    }
-
     /**
      * @param key Key.
      * @return {@code True} if mapped.
      */
-    private boolean map(KeyCacheObject key, boolean forceKeys) {
+    private boolean reservePartition(KeyCacheObject key) {
         try {
             int keyPart = cctx.affinity().partition(key);
 
-            if (cctx.mvccEnabled()) {
-                boolean noOwners = cctx.topology().owners(keyPart, topVer).isEmpty();
-
-                // Force key request is disabled for MVCC. So if there are no partition owners for the given key
-                // (we have a not strict partition loss policy if we've got here) we need to set flag forceKeys to true
-                // to avoid useless remapping to other non-owning partitions. For non-mvcc caches the force key request
-                // is also useless in the such situations, so the same flow is here: allegedly we've made a force key
-                // request with no results and therefore forceKeys flag may be set to true here.
-                if (noOwners)
-                    forceKeys = true;
-            }
-
             GridDhtLocalPartition part = topVer.topologyVersion() > 0 ?
                 cache().topology().localPartition(keyPart, topVer, true) :
                 cache().topology().localPartition(keyPart);
```

Comment:
I don't think we should create a partition on a get in any circumstances.
This should be simplified to GridDhtLocalPartition part = cache().topology().localPartition(keyPart);